File: Java/Binary Search/Medium/FindSmallestDivisor.java
// 1283
class Solution {
    public int smallestDivisor(int[] nums, int threshold) {
        int max = 0;
        for (int num : nums) {
            max = Math.max(max, num);
        }

        int low = 1, high = max;
        while (low < high) {
            int mid = low + (high - low) / 2;
            if (isValid(nums, mid, threshold)) {
                high = mid; // divisor works, try smaller
            } else {
                low = mid + 1; // divisor too small, increase
            }
        }
        return low;
    }

    private boolean isValid(int[] nums, int divisor, int threshold) {
        int sum = 0;
        for (int num : nums) {
            sum += (num + divisor - 1) / divisor; // ceil(num/divisor)
            if (sum > threshold) return false; // prune early
        }
        return true;
    }
}

File: Binary Search/FindSmallestDivisor.py
from typing import List
import math
# 1283
class Solution:
    def smallestDivisor(self, nums: List[int], threshold: int) -> int:
        
        lo = 1
        hi = max(nums)
        res = 1
        while lo <= hi:
            mid = (lo + hi) // 2
            total = 0
            for num in nums:
                val = math.ceil(num / mid)
                total += val
            if total <= threshold:
                res = mid
                hi = mid - 1
            else:
                lo = mid + 1
        return res
